extensions/amp-story/0.1/amp-story-auto-ads.js

```diff
@@ -29,42 +34,151 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
   constructor(element) {
     super(element);
 
-    /** @private {?AmpStory} */
+    /** @private {?./amp-story.AmpStory} */
     this.ampStory_ = null;
+
+    /** @private {?./navigation-state.NavigationState} */
+    this.navigationState_ = null;
+
+    /** @private {number} */
+    this.interactions_ = 0;
+
+    /** @private {!Array} */
+    this.adElements_ = [];
+
+    /** @private {number} */
+    this.adIdPointer_ = 0;
+
+    /** @private {?Array} */
+    this.pages_ = null;
   }
 
   /** @override */
   buildCallback() {
-    const ampStory = this.element.parentElement;
-    user().assert(ampStory.tagName === 'AMP-STORY',
+    const ampStoryElement = this.element.parentElement;
+    user().assert(ampStoryElement.tagName === 'AMP-STORY',
         `<${TAG}> should be child of <amp-story>`);
-    this.ampStory_ = ampStory;
-  }
+
+    ampStoryElement.getImpl().then(impl => {
```

Comment:
Instead of doing this, I think you can listen on the `amp-story` element for `NEXT_PAGE`, `PREVIOUS_PAGE`, and/or `SWITCH_PAGE` events.  Then, you don't need the impl or the `NavigationState`.

Reply:
ok I will give this a shot

Reply:
@newmuis I actually think NavigationState is helpful here as it wraps the raw events and gives a semantic view of the data.

I agree that this extension should not depend on amp-story. eventually, we should have a global NavigatorService that abstract out the navigation related logic, together with a PageManagerService.

Reply:
left this after our conversation